block/manager.go

```diff
@@ -101,11 +101,23 @@ type Manager struct {
 
 // getInitialState tries to load lastState from Store, and if it's not available it reads GenesisDoc.
 func getInitialState(store store.Store, genesis *cmtypes.GenesisDoc) (types.State, error) {
+	b, err := store.GetBlock(uint64(genesis.InitialHeight))
+	if err != nil {
+		// we've never seen a block at InitialHeight.
+		// assume we're syncing from a trusted snapshot
+		return types.NewFromGenesisDoc(genesis)
+	}
+	if !bytes.Equal(b.SignedHeader.AppHash, genesis.AppHash.Bytes()) {
+		// the user-supplied genesis overrides a previously-seen block
+		// must be a hard-fork.
+		return types.NewFromGenesisDoc(genesis)
+	}
 	s, err := store.GetState()
```

Comment:
can you explain why we need additional logic to support hard-fork? if there is a trusted snapshot, the `store.GetState` would give you the latest trusted block state from the loaded snapshot right? because, the snapshot will always have the updated state.

Reply:
what i mean is, the hard-fork itself is a state with 1 or more blocks different from other state (non-hard-fork).

Reply:
are you saying we don't need this because hard-forks will replace the store?

Reply:
yes